packages/bug-detectors/internal/eval.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Code Intelligence GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+	getJazzerJsGlobal,
+	guideTowardsContainment,
+	reportAndThrowFinding,
+	setJazzerJsGlobal,
+} from "@jazzer.js/core";
+import { callSiteId } from "@jazzer.js/hooking";
+import * as vm from "vm";
+
+const targetString = "jaz_zer";
+
+const originalEval = eval;
+const originalFunction = Function;
+
+globalThis.eval = hookEval(originalEval);
+// @ts-ignore
+globalThis.Function = hookFunction(originalFunction);
+
+if (getJazzerJsGlobal("vmContext")) {
+	const vmContext = getJazzerJsGlobal("vmContext") as vm.Context;
+	// The VM should use eval from its own context, and not the global eval.
+	vmContext.eval = hookEval(vm.runInContext("eval", vmContext));
+	vmContext.Function = hookFunction(vm.runInContext("Function", vmContext));
+}
+
+function hookEval(originalEval: typeof eval) {
```

Comment:
Why not use `registerBeforeHook`, as for example the command injection detector?